### ISSUE TITLE
GRPO 训练过程中突然OOM

### ISSUE THREAD

**Issue by SLKAlgs** [modelscope/ms-swift#6398]
4*A100(80G)使用grpo训练，前期显存每张卡占用约27G，但是训练过程中会突然报错OOM，训练脚本如下：
export MASTER_PORT=29501
MAX_PIXELS=602112 \
NPROC_PER_NODE=4 \
swift rlhf \
    --rlhf_type grpo \
    --model /home/qid/lgs/lgs-foodkg/ms-swift/output_kg/v1-20251027-102007/checkpoint-1875-merged \
    --train_type lora \
    --dataset '/home/qid/lgs/lgs-foodkg/ms-swift/data/grpo_train_entities_full.json' \
    --use_vllm true \
    --external_plugins examples/train/grpo/plugin/plugin.py \
    --vllm_mode colocate \
    --vllm_gpu_memory_utilization 0.5 \
    --vllm_tensor_parallel_size 4 \
    --torch_dtype bfloat16 \
    --system /home/qid/lgs/lgs-foodkg/ms-swift/examples/train/grpo/plugin/kg_prompt.txt \
    --num_train_epochs 1 \
    --per_device_train_batch_size 1 \
    --per_device_eval_batch_size 1 \
    --split_dataset_ratio 0 \
    --learning_rate 5e-6 \
    --save_total_limit 2 \
    --logging_steps 1 \
    --output_dir output_grpo \
    --gradient_accumulation_steps 4 \
    --warmup_ratio 0.05 \
    --dataloader_num_workers 4 \
    --max_completion_length 4096 \
    --reward_funcs entity relation format \
    --reward_weights 1 1 0.1 \
    --num_generations 8 \
    --sleep_level 1 \
    --save_steps 500 \
    --temperature 1.0 \
    --top_p 0.85 \
    --gradient_checkpointing_kwargs '{"use_reentrant": true}' \
    --max_length 8192 \
    --lora_rank 128 \
    --lora_alpha 256 \
    --attn_impl flash_attn \
    --deepspeed zero3 \
    --top_k 50 \
    --report_to wandb
求问这是为什么 谢谢！

**Comment by hjh0119** [comment]
https://swift.readthedocs.io/zh-cn/latest/Instruction/GRPO/GetStarted/GRPO.html#colocate-internal-mode

**Comment by SLKAlgs** [comment]
> https://swift.readthedocs.io/zh-cn/latest/Instruction/GRPO/GetStarted/GRPO.html#colocate-internal-mode

谢谢您的回复！还有一个问题，我设置了三个reward函数（包括format），但是另外两个reward一直极其不稳定，没有上升趋势，模型没有学习到，如图所示，请问可能是什么原因呢？我的reward函数逻辑如下，即计算预测和gt之间的F1分数，若大于阈值则加分：
class KGEntityReward(ORM):

    def __init__(self, threshold: float = 0.25):
        self.threshold = threshold

    def __call__(self, completions, **kwargs) -> List[float]:
        rewards: List[float] = []
        references = _extract_reference_texts(kwargs, completions)
        _debug_logged['entity'] = True
        for idx, (completion, reference) in enumerate(zip(completions, references)):
            f1 = 0.0
            try:
                pred_entities_block, pred_rel_block = _split_sections(completion)
                gt_entities_block, gt_rel_block = _split_sections(reference)
                if not (pred_entities_block and pred_rel_block and gt_entities_block and gt_rel_block):
                    rewards.append(0.0)
                    logger.info('[entity][%s] f1=%.4f 加分=%s (缺少实体或关系片段)', idx, f1, '否')
                    continue

                pred_entities = parse_entities(pred_entities_block)
                gt_entities = parse_entities(gt_entities_block)
                pred_lookup = build_entity_lookup(pred_entities)
                gt_lookup = build_entity_lookup(gt_entities)

                pred_triples = parse_triples(pred_rel_block)
                gt_triples = parse_triples(gt_rel_block)
                if not pred_triples or not gt_triples:
                    rewards.append(0.0)
                    logger.info('[entity][%s] f1=%.4f 加分=%s (缺少三元组)', idx, f1, '否')
                    continue

                pred_counter = count_entities(pred_triples, pred_lookup, strict=False)
                gt_counter = count_entities(gt_triples, gt_lookup, strict=False)
                f1 = _compute_f1_from_counters(pred_counter, gt_counter)
                reward = 1.0 if f1 > self.threshold else 0.0
                rewards.append(reward)
                logger.info('[entity][%s] predicted_entities=%s', idx, _format_entities(pred_entities))
                logger.info('[entity][%s] gt_entities=%s', idx, _format_entities(gt_entities))
                logger.info('[entity][%s] f1=%.4f 加分=%s', idx, f1, '是' if reward > 0 else '否')
            except Exception as exc:
                rewards.append(0.0)
                logger.warning('[entity][%s] 解析失败 %s，f1=%.4f 加分=否', idx, exc, f1)
        return rewards


<img width="1568" height="692" alt="Image" src="https://github.com/user-attachments/assets/d5436ace-101d-41cb-ade5-14429ef17e5b" />

**Comment by hjh0119** [comment]
maybe you can try full parameter training

**Comment by aLittleTinyBoy** [comment]
你好，我也遇到了在GRPO训练中突然OOM的问题，你是用手册中哪条指令解决的？是--sleep_level 1吗？
